src/abi/io.rs

```diff
@@ -0,0 +1,18 @@
+use object::ProcessRef;
+use nebulet_derive::nebulet_abi;
+use alloc::String;
+
+#[nebulet_abi]
+pub fn print(buffer_offset: u32, buffer_size: u32, process: &ProcessRef) {
+    let instance = process.instance().read();
+    let memory = &instance.memories[0];
+
+    if let Some(buf) = memory.carve_slice(buffer_offset, buffer_size) {
+        // TODO: Make happy path not allocate
+        let s = String::from_utf8_lossy(buf);
```

Comment:
Looks like `String::from_utf8_lossy` only allocates when the input is not valid utf-8, so this should be alright, unless you think we should reject invalid utf-8.

Reply:
Uh, it has to allocate. I'm passing in a `&[u8]` and getting a `Vec<u8>` (well `String`, but same thing really).

I don't think we should reject invalid utf-8

Reply:
Nope, `String::from_ut8_lossy` returns a `Cow<str>`, which allocates on write, but otherwise keeps the same buffer.

Reply:
Oh, TIL. Will remove that todo then